pkg/executor/join/hash_join_base.go

```diff
@@ -0,0 +1,359 @@
+	for i := uint(0); i < concurrency; i++ {
+		fetcher.probeResultChs[i] = make(chan *chunk.Chunk, 1)
+	}
+	// fetcher.probeChkResourceCh is for transmitting the used ProbeSideExec chunks from
+	// join workers to ProbeSideExec worker.
+	fetcher.probeChkResourceCh = make(chan *probeChkResource, concurrency)
+	for i := uint(0); i < concurrency; i++ {
+		fetcher.probeChkResourceCh <- &probeChkResource{
+			chk:  exec.NewFirstChunk(fetcher.ProbeSideExec),
+			dest: fetcher.probeResultChs[i],
+		}
+	}
+	fetcher.joinResultChannel = joinResultChannel
+}
+
+func (fetcher *probeSideTupleFetcherBase) handleProbeSideFetcherPanic(r any) {
+	for i := range fetcher.probeResultChs {
+		close(fetcher.probeResultChs[i])
+	}
+	if r != nil {
+		fetcher.joinResultChannel <- &hashjoinWorkerResult{err: util.GetRecoverError(r)}
+	}
+}
+
+type isBuildSideEmpty func() bool
+
+func wait4BuildSide(isBuildEmpty isBuildSideEmpty, canSkipIfBuildEmpty bool, hashJoinCtx *hashJoinCtxBase) (skipProbe bool, err error) {
+	select {
+	case <-hashJoinCtx.closeCh:
+		return true, nil
+	case err := <-hashJoinCtx.buildFinished:
+		if err != nil {
+			return false, err
+		}
+	}
+	if isBuildEmpty() && canSkipIfBuildEmpty {
+		return true, nil
+	}
+	return false, nil
+}
+
+// fetchProbeSideChunks get chunks from fetches chunks from the big table in a background goroutine
+// and sends the chunks to multiple channels which will be read by multiple join workers.
+func (fetcher *probeSideTupleFetcherBase) fetchProbeSideChunks(ctx context.Context, maxChunkSize int, isBuildEmpty isBuildSideEmpty, canSkipIfBuildEmpty, needScanAfterProbeDone, shouldLimitProbeFetchSize bool, hashJoinCtx *hashJoinCtxBase) {
+	hasWaitedForBuild := false
+	for {
+		if hashJoinCtx.finished.Load() {
+			return
+		}
+
+		var probeSideResource *probeChkResource
+		var ok bool
+		select {
+		case <-hashJoinCtx.closeCh:
+			return
+		case probeSideResource, ok = <-fetcher.probeChkResourceCh:
+			if !ok {
+				return
+			}
+		}
+		probeSideResult := probeSideResource.chk
+		if shouldLimitProbeFetchSize {
+			required := int(atomic.LoadInt64(&fetcher.requiredRows))
+			probeSideResult.SetRequiredRows(required, maxChunkSize)
+		}
+		err := exec.Next(ctx, fetcher.ProbeSideExec, probeSideResult)
+		failpoint.Inject("ConsumeRandomPanic", nil)
+		if err != nil {
+			hashJoinCtx.joinResultCh <- &hashjoinWorkerResult{
+				err: err,
+			}
+			return
+		}
+		if !hasWaitedForBuild {
+			failpoint.Inject("issue30289", func(val failpoint.Value) {
+				if val.(bool) {
+					probeSideResult.Reset()
+				}
+			})
+			skipProbe, buildErr := wait4BuildSide(isBuildEmpty, canSkipIfBuildEmpty, hashJoinCtx)
+			if buildErr != nil {
+				hashJoinCtx.joinResultCh <- &hashjoinWorkerResult{
+					err: buildErr,
+				}
+				return
+			} else if skipProbe {
+				// stop probe
+				if !needScanAfterProbeDone {
+					hashJoinCtx.finished.Store(true)
+				}
+				return
+			}
+			hasWaitedForBuild = true
+		}
```

Comment:
Maybe we can encapsulate these codes into a function named like `tryToWaitForbuild`.

Reply:
Put some code into `wait4BuildSide`